> r? me
> 
> We already have a `SuggestBoxingForReturnImplTrait` suggestion.
> 
That is was used in the pr.
> Please use it instead. Or ideally, figure out why it's not being emitted in some cases, like [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=0f2f61ef15ed2d71848dc12b6a6e2f86)
> 
> ```rust
> // suggestion emitted
> fn works() -> impl std::fmt::Display {
>     match 13 {
>         0 => 0i32,
>         _ => 1u32,
>     }
> }
> 
> // suggestion not given
> fn oh_no() -> impl std::fmt::Display {
>     match 13 {
>         0 => return 0i32,
>         _ => 1u32,
>     }
> }
> ```
> 
> ```
> error[E0308]: `match` arms have incompatible types
>  --> src/lib.rs:5:14
>   |
> 3 | /     match 13 {
> 4 | |         0 => 0i32,
>   | |              ---- this is found to be of type `i32`
> 5 | |         _ => 1u32,
>   | |              ^^^^ expected `i32`, found `u32`
> 6 | |     }
>   | |_____- `match` arms have incompatible types
>   |
> help: you could change the return type to be a boxed trait object
>   |
> 2 - fn works() -> impl std::fmt::Display {
> 2 + fn works() -> Box<dyn std::fmt::Display> {
>   |
> help: if you change the return type to expect trait objects, box the returned expressions
>   |
> 4 ~         0 => Box::new(0i32),
> 5 ~         _ => Box::new(1u32),
>   |
> help: change the type of the numeric literal from `u32` to `i32`
>   |
> 5 -         _ => 1u32,
> 5 +         _ => 1i32,
>   |
> 
> error[E0308]: mismatched types
>   --> src/lib.rs:13:14
>    |
> 10 | fn oh_no() -> impl std::fmt::Display {
>    |               ---------------------- expected a single type implementing `std::fmt::Display` because of return type
> 11 |     match 13 {
> 12 |         0 => return 0i32,
>    |                     ---- return type resolved to be `i32`
> 13 |         _ => 1u32,
>    |              ^^^^ expected `i32`, found `u32`
>    |
> help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
>    |
> 14 |     }.try_into().unwrap()
>    |      ++++++++++++++++++++
> 
> For more information about this error, try `rustc --explain E0308`.
> ```
> 
> _[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/155682/92c7010294007f9bb819f0ddd9d9e9b5ccd5714a..360e5d19487a922d0c964186f4bd247259519e23)_

Those two are different errors and are emitted at different locations. While we could make the second case suggest `Box::new()` for the return type, it is cleaner to let the compiler guide the user (if the user adds `Box<dyn std::fmt::Display>`, the compiler will then suggest `Box::new()` in the return expressions). The location where the error is emitted makes it a bit harder to implement, and it could become messy or overly verbose if there are multiple return points, such as in an `if-else` chain.

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -7,6 +7,7 @@ use rustc_ast::util::parser::ExprPrecedence;
 use rustc_data_structures::packed::Pu128;
 use rustc_errors::{Applicability, Diag, MultiSpan, listify, msg};
 use rustc_hir::def::{CtorKind, CtorOf, DefKind, Res};
+use rustc_hir::intravisit::Visitor;
 use rustc_hir::lang_items::LangItem;
 use rustc_hir::{
     self as hir, Arm, CoroutineDesugaring, CoroutineKind, CoroutineSource, Expr, ExprKind,
@@ -26,6 +27,7 @@ use rustc_session::errors::ExprParenthesesNeeded;
 use rustc_span::{ExpnKind, Ident, MacroKind, Span, Spanned, Symbol, sym};
 use rustc_trait_selection::error_reporting::InferCtxtErrorExt;
 use rustc_trait_selection::error_reporting::traits::DefIdOrName;
+use rustc_trait_selection::error_reporting::traits::suggestions::ReturnsVisitor;
 use rustc_trait_selection::infer::InferCtxtExt;
 use rustc_trait_selection::traits;
 use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt as _;
@@ -970,7 +972,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             end_sp: hir_ty.span.shrink_to_hi(),
                         });
 
-                        err.note("if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`");
+                        let body = self.tcx.hir_body_owned_by(fn_id);
+                        let mut visitor = ReturnsVisitor::default();
+                        visitor.visit_body(&body);
+
+                        if !visitor.returns.is_empty() {
+                            let starts: Vec<Span> = visitor
+                                .returns
+                                .iter()
+                                .filter(|expr| expr.span.can_be_used_for_suggestions())
+                                .map(|expr| expr.span.shrink_to_lo())
+                                .collect();
+                            let ends: Vec<Span> = visitor
+                                .returns
+                                .iter()
+                                .filter(|expr| expr.span.can_be_used_for_suggestions())
+                                .map(|expr| expr.span.shrink_to_hi())
+                                .collect();
+
+                            if !starts.is_empty() {
+                                err.subdiagnostic(SuggestBoxingForReturnImplTrait::BoxReturnExpr {
+                                    starts,
+                                    ends,
+                                });
+                            }
+                        }
                     }
 
                     self.try_suggest_return_impl_trait(err, expected, found, fn_id);

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -32,7 +32,7 @@ use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt as _
 use tracing::{debug, instrument};
 
 use super::FnCtxt;
-use crate::errors;
+use crate::errors::{self, SuggestBoxingForReturnImplTrait};
 use crate::fn_ctxt::rustc_span::BytePos;
 use crate::method::probe;
 use crate::method::probe::{IsSuggestion, Mode, ProbeScope};
@@ -961,6 +961,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             ret_coercion_span,
                             format!("return type resolved to be `{expected_name}`"),
                         );
+                    }
+
+                    let trait_def_id = trait_ref.trait_ref.path.res.def_id();
+                    if self.tcx.is_dyn_compatible(trait_def_id) {
+                        err.subdiagnostic(SuggestBoxingForReturnImplTrait::ChangeReturnType {
+                            start_sp: hir_ty.span.with_hi(hir_ty.span.lo() + BytePos(4)),
+                            end_sp: hir_ty.span.shrink_to_hi(),
+                        });
+
+                        err.note("if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`");
                     }
 
                     self.try_suggest_return_impl_trait(err, expected, found, fn_id);

--- a/tests/ui/impl-trait/dyn-incompatible-trait-in-return-position-impl-trait.stderr
+++ b/tests/ui/impl-trait/dyn-incompatible-trait-in-return-position-impl-trait.stderr
@@ -22,11 +22,16 @@ LL |     }
 LL |     B
    |     ^ expected `A`, found `B`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn cat() -> impl DynCompatible {
 LL + fn cat() -> Box<dyn DynCompatible> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~         return Box::new(A);
+LL |     }
+LL ~     Box::new(B)
    |
 
 error: aborting due to 2 previous errors

--- a/tests/ui/impl-trait/dyn-incompatible-trait-in-return-position-impl-trait.stderr
+++ b/tests/ui/impl-trait/dyn-incompatible-trait-in-return-position-impl-trait.stderr
@@ -21,6 +21,13 @@ LL |         return A;
 LL |     }
 LL |     B
    |     ^ expected `A`, found `B`
+   |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn cat() -> impl DynCompatible {
+LL + fn cat() -> Box<dyn DynCompatible> {
+   |
 
 error: aborting due to 2 previous errors
 

--- a/tests/ui/impl-trait/point-to-type-err-cause-on-impl-trait-return.stderr
+++ b/tests/ui/impl-trait/point-to-type-err-cause-on-impl-trait-return.stderr
@@ -72,6 +72,12 @@ LL |     }
 LL |     1u32
    |     ^^^^ expected `i32`, found `u32`
    |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn foo() -> impl std::fmt::Display {
+LL + fn foo() -> Box<dyn std::fmt::Display> {
+   |
 help: change the type of the numeric literal from `u32` to `i32`
    |
 LL -     1u32
@@ -90,6 +96,12 @@ LL |     } else {
 LL |         return 1u32;
    |                ^^^^ expected `i32`, found `u32`
    |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn bar() -> impl std::fmt::Display {
+LL + fn bar() -> Box<dyn std::fmt::Display> {
+   |
 help: change the type of the numeric literal from `u32` to `i32`
    |
 LL -         return 1u32;
@@ -107,6 +119,12 @@ LL |         return 0i32;
 LL |     } else {
 LL |         1u32
    |         ^^^^ expected `i32`, found `u32`
+   |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn baz() -> impl std::fmt::Display {
+LL + fn baz() -> Box<dyn std::fmt::Display> {
    |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |
@@ -153,6 +171,12 @@ LL |         0 => return 0i32,
 LL |         _ => 1u32,
    |              ^^^^ expected `i32`, found `u32`
    |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn bat() -> impl std::fmt::Display {
+LL + fn bat() -> Box<dyn std::fmt::Display> {
+   |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |
 LL |     }.try_into().unwrap()
@@ -171,6 +195,12 @@ LL | |         _ => 2u32,
 LL | |     }
    | |_____^ expected `i32`, found `u32`
    |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn can() -> impl std::fmt::Display {
+LL + fn can() -> Box<dyn std::fmt::Display> {
+   |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |
 LL |     }.try_into().unwrap()
@@ -187,6 +217,12 @@ LL |             return 0i32;
 ...
 LL |             1u32
    |             ^^^^ expected `i32`, found `u32`
+   |
+   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
+help: you could change the return type to be a boxed trait object
+   |
+LL - fn cat() -> impl std::fmt::Display {
+LL + fn cat() -> Box<dyn std::fmt::Display> {
    |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |

--- a/tests/ui/impl-trait/point-to-type-err-cause-on-impl-trait-return.stderr
+++ b/tests/ui/impl-trait/point-to-type-err-cause-on-impl-trait-return.stderr
@@ -72,11 +72,16 @@ LL |     }
 LL |     1u32
    |     ^^^^ expected `i32`, found `u32`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn foo() -> impl std::fmt::Display {
 LL + fn foo() -> Box<dyn std::fmt::Display> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~         return Box::new(0i32);
+LL |     }
+LL ~     Box::new(1u32)
    |
 help: change the type of the numeric literal from `u32` to `i32`
    |
@@ -96,11 +101,16 @@ LL |     } else {
 LL |         return 1u32;
    |                ^^^^ expected `i32`, found `u32`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn bar() -> impl std::fmt::Display {
 LL + fn bar() -> Box<dyn std::fmt::Display> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~         return Box::new(0i32);
+LL |     } else {
+LL ~         return Box::new(1u32);
    |
 help: change the type of the numeric literal from `u32` to `i32`
    |
@@ -120,11 +130,16 @@ LL |     } else {
 LL |         1u32
    |         ^^^^ expected `i32`, found `u32`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn baz() -> impl std::fmt::Display {
 LL + fn baz() -> Box<dyn std::fmt::Display> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~         return Box::new(0i32);
+LL |     } else {
+LL ~         Box::new(1u32)
    |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |
@@ -171,11 +186,15 @@ LL |         0 => return 0i32,
 LL |         _ => 1u32,
    |              ^^^^ expected `i32`, found `u32`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn bat() -> impl std::fmt::Display {
 LL + fn bat() -> Box<dyn std::fmt::Display> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~         0 => return Box::new(0i32),
+LL ~         _ => Box::new(1u32),
    |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |
@@ -195,11 +214,16 @@ LL | |         _ => 2u32,
 LL | |     }
    | |_____^ expected `i32`, found `u32`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn can() -> impl std::fmt::Display {
 LL + fn can() -> Box<dyn std::fmt::Display> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~         0 => return Box::new(0i32),
+LL ~         1 => Box::new(1u32),
+LL ~         _ => Box::new(2u32),
    |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |
@@ -218,11 +242,17 @@ LL |             return 0i32;
 LL |             1u32
    |             ^^^^ expected `i32`, found `u32`
    |
-   = note: if you change the return type to expect trait objects, you'll need to wrap the returned values in `Box::new()`
 help: you could change the return type to be a boxed trait object
    |
 LL - fn cat() -> impl std::fmt::Display {
 LL + fn cat() -> Box<dyn std::fmt::Display> {
+   |
+help: if you change the return type to expect trait objects, box the returned expressions
+   |
+LL ~             return Box::new(0i32);
+LL |         }
+LL |         _ => {
+LL ~             Box::new(1u32)
    |
 help: you can convert a `u32` to an `i32` and panic if the converted value doesn't fit
    |

